kroxylicious-filters/kroxylicious-oauthbearer-validation/src/main/java/io/kroxylicious/proxy/filter/oauthbearer/OauthBearerValidationFilter.java

```diff
@@ -0,0 +1,212 @@
+                    .shortCircuitResponse(new SaslHandshakeResponseData().setErrorCode(UNKNOWN_SERVER_ERROR.code()))
+                    .withCloseConnection()
+                    .completed();
+        }
+        return context.forwardRequest(header, request);
+    }
+
+    @Override
+    public CompletionStage<RequestFilterResult> onSaslAuthenticateRequest(short apiVersion, RequestHeaderData header,
+                                                                          SaslAuthenticateRequestData request, FilterContext context) {
+        if (!validateAuthentication) {
+            // client is already authenticated or is not using OAUTHBEARER mechanism, we can forward the request to cluster
+            return context.forwardRequest(header, request);
+        }
+        else {
+            return authenticate(request.authBytes())
+                    .thenCompose(bytes -> context.forwardRequest(header, request))
+                    .exceptionallyCompose(e -> {
+                        if (e.getCause() instanceof SaslAuthenticationException) {
+                            SaslAuthenticateResponseData failedResponse = new SaslAuthenticateResponseData()
+                                    .setErrorCode(SASL_AUTHENTICATION_FAILED.code())
+                                    .setErrorMessage(e.getMessage())
+                                    .setAuthBytes(request.authBytes());
+                            LOGGER.debug("SASL Authentication failed : {}", e.getMessage(), e);
+                            return context.requestFilterResultBuilder().shortCircuitResponse(failedResponse).withCloseConnection().completed();
+                        }
+                        else {
+                            LOGGER.debug("SASL error : {}", e.getMessage(), e);
+                            return context.requestFilterResultBuilder()
+                                    .shortCircuitResponse(
+                                            new SaslAuthenticateResponseData()
+                                                    .setErrorCode(UNKNOWN_SERVER_ERROR.code())
+                                                    .setAuthBytes(request.authBytes()))
+                                    .withCloseConnection()
+                                    .completed();
+                        }
+                    });
+        }
+    }
+
+    @Override
+    public CompletionStage<ResponseFilterResult> onSaslAuthenticateResponse(short apiVersion, ResponseHeaderData header,
+                                                                            SaslAuthenticateResponseData response, FilterContext context) {
+        if (response.errorCode() == NONE.code()) {
+            this.validateAuthentication = false;
+        }
+        return context.forwardResponse(header, response);
+    }
+
+    private CompletionStage<byte[]> authenticate(byte[] authBytes) {
+        String digest;
+        try {
+            digest = digestBytes(authBytes);
```

Comment:
Likewise `digestBytes` tells us something about the implementation of the rate limiter key, but at this point in the code we only depend on its `equality` and `hashCode`, not the fact that it's a digest per se.